proposals/compound-extensions.md

```diff
@@ -0,0 +1,380 @@
+   ```kotlin
+   html {
+        body {
+            operator fun String.unaryMinus() {
+                underline {
+                    +this@unaryMinus
+                }
+            }
+            div {
+                -"Customer"
+                customer(customer)
+
+                -"Orders"
+                orders(orders)
+            }
+        }
+    }
+    ```
+
+    However, the developer is required to modify `Body` class to allow the unary `-` to be
+    used outside its current local scope. This is problematic if the HTML DSL is a third-party
+    library.
+
+    Compound extensions allows the original unary `+` operator to be introduced as an extension
+    function such as,
+
+    ```kotlin
+    operator fun Body.String.unaryPlus() = escape(this)
+    ```
+
+    therefore allowing unary `-` to be provided generally by extension of `Body`,
+
+    ```kotlin
+    operator fun Body.String.unaryMinus() = underline { +this@unaryMinus }
+    ```
+
+## Proposal
+
+### Syntax
+
+No new syntax is required. Extension declarations that are currently invalid are given
+a valid interpretation.
+
+For example, in the above examples,
+
+```kotlin
+operator fun Body.String.unaryPlus() = escape(this)
+```
+
+is syntactically valid but reports that `String` is not valid as there is no nested
+`String` class of `Body`.
+
+The type expression syntax is similarly interpreted. For example the type of,
+
+```kotlin
+fun A.B.C.someMethod(v: Int): Int = ...
+```
+
+would be written, `A.B.C.(Int) -> Int`.
+
+#### Alternate syntax
+
+Syntax is always contentious. The problem with the primary proposal is the ambiguity it has
+with namespaces. Other syntaxes to consider are:
+
+##### Using parenthesis
+```kotlin
+fun (A, B, C).someMethod(v: Int): Int = ...
+
+val a: (A,B,C).(v: Int)->Int = ...
+```
+
+The types are unambiguous such that `fun (A.B, C).someMethod(v: Int)`, the `A.B` is
+unambiguously a reference to the nested type `B` in `A`.
+
+The type expression syntax would be similarly extended to allow `(A.B, C).() -> Unit`.
+
+However, the type expression syntax is ambiguous at the `(` and it is not until the `.` is
+seen after the initial closing `)`.
+
+#### Using brackets
+```kotlin
+fun [A, B, C].someMethod(v: Int): Int = ...
+
+val a: [A,B,C].(Int)->Int = ...
+```
+
+The types are similarly unambiguous without introducing the same syntactic ambiguity. However,
+this require introducing a use for `[` in a type expression that might be better reserved by
+a feature more widely leveraged such as tuples.
+
+#### Using a pseudo-keyword
+
+```kotlin
+extension fun (A, B, C).someMethod(v: Int): Int = ...
+val a: extension (A, B, C).(Int)->Int = ...
+```
+
+Using a pseudo keyword avoids the ambiguity of the `(` in the type expression. A pseudo-
+keyword could also be used with the `[`...`]` syntax to leave a unadorned `[`...`]` to
+mean a tuple in the future.
+
+### Matching rules
+
+1. **Create an ordered list of implicit receivers** in scope with types
```

Comment:
I would like to see support for explicit receivers for multi-receiver extensions.  Existing extensions can be called with an explicit or implicit receiver, and I think there are cases where explicit receivers could be important for multi-receiver extensions.

As noted below, receivers may be "explicitly" provided with the following syntax, but it extremely verbose and not obvious:
```
with (a) {
    with (b) {
        with (c) {
            doSomething()
        }
    }
}
```

Reply:
What about:
```
with (a,b,c) { doSomething() }
```
Where the receivers match the order of the receivers of `doSomething`